model.py

```diff
@@ -0,0 +1,141 @@
+
+    def __init__(self,
+                 c_in: int,
+                 c_out: int,
+                 stride: Union[int, Tuple[int, int, int]],
+                 use_bn: bool = True,
+                 use_up: bool = False,
+                 lrelu_eps: float = 0.2,
+                 ):
+        """
+
+        Args:
+            c_in: Number of input channels.
+            c_out: Number of output channels.
+            stride: Convolution stride.
+            use_bn: Whether to use batch normalization. (nn.BatchNorm3d)
+            use_up: Whether to use trilinear upsampling (nn.Upsample)
+            lrelu_eps: nn.LeakyRelu slope parameter.
+                WARN(ycho): defaults to 0.2 (according to the paper),
+                but beware that the author's implementation seems to use 1e-2.
+        """
+        super().__init__()
+
+        layers: List[nn.Module] = []
+        layers.append(nn.Conv3d(
+            c_in, c_out, (3, 3, 3),
+            stride, padding=1, bias=False))
+        if use_bn:
+            layers.append(nn.BatchNorm3d(c_out))
+        layers.append(nn.LeakyReLU(lrelu_eps, inplace=True))
+        if use_up:
+            layers.append(
+                nn.Upsample(
+                    scale_factor=2,
+                    mode='trilinear',
+                    align_corners=True))
+        self.layers = nn.Sequential(*layers)
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        return self.layers(x)
+
+
+class ConvResBlock(nn.Module):
+    """Convolutional residual block, x'=x+f(x)."""
+
+    def __init__(self, c_in: int, c_out: int,
+                 lrelu_eps: float = 1e-2):
+        """
+
+        Args:
+            c_in: Number of input channels.
+            c_out: Number of output channels.
+            lrelu_eps: nn.LeakyRelu slope parameter.
+        """
+        super().__init__()
+        layers: List[nn.Module] = []
+        layers.append(nn.Conv3d(c_in, c_out, 3, padding=1, bias=False))
+        layers.append(nn.BatchNorm3d(c_out))
+        layers.append(nn.LeakyReLU(lrelu_eps, inplace=True))
+        layers.append(nn.Conv3d(c_out, c_out, 3, padding=1, bias=False))
+        layers.append(nn.BatchNorm3d(c_out))
+        self.layers = nn.Sequential(*layers)
+
+    def forward(self, x: th.Tensor) -> th.Tensor:
+        return x + self.layers(x)
+
+
+class SceneEncoder(nn.Module):
+    """DSR-Net Scene Encoder.
+
+    Produces current latent-state vector from TSDF-formatted
+    observations and the previous state.
+    """
+
+    def __init__(self):
+        super().__init__()
+        # TODO(ycho): cleanup ugly syntax
+
+        # Encoder part
+        # NOTE(ycho): in author's implementation, number of input channels
+        # are `12` (8+1+3) due to undocumented inclusion of
+        # meshgrid "coordinate features" (const) rather than `9` (8+1).
+        enc_channels: Tuple[int, ...] = (
+            9,
+            16, 32, 32, 32,
+            32, 64, 64, 64, 64,
+            128
+        )
```

Comment:
Can we somehow regroup this like `((9, 16), (16, 32, 32, 32, 32)...` or generate that from `enc_channels` and `(0, 1, 5, 9)`?

Reply:
I guess we can regroup, but is there a reason you'd prefer that way?

Reply:
On the second thought, my first suggestion looks redundant. I think simply extracting `(0, 1, 5, 9)` to a variable/constant will improve the code.

Reply:
I'm not quite sure what your suggestion will end up looking like.
So, I'll merge the code as-is for now.
However, feel free to modify `SceneEncoder.__init__()` as you like, in a subsequent PR.